lib/state.js

```diff
@@ -0,0 +1,512 @@
+   * function does *not* run the queries internally, rather uses the database
+   * connection parameters value that this State's pg-pool was configured with
+   * to run the script using the command line 'psql' program.  It chooses the
+   * first 'psql' program in the system's path to run the script.
+   * 
+   * This is done because I'd like to keep the database definition queries as
+   * standard sql scripts, while allowing their use in the setup and teardown
+   * sections of the tests
+   */
+  async _runScript(script) {
+    let args = [];
+    let o = this._pgpool.options; // cuts down on verbosity
+    if (o.host) {
+      args.push(`--host=${o.host}`);
+    }
+    if (o.port) {
+      args.push(`--port=${o.port}`);
+    }
+    if (o.database) {
+      args.push(`--dbname=${o.database}`);
+    }
+    if (o.user) {
+      args.push(`--username=${o.user}`);
+    }
+    args.push('--file=' + path.normalize(path.join(__dirname, '..', 'sql', script)));
+
+    let env = {};
+    if (o.password) {
+      env.PGPASSWORD = o.password;
+    }
+
+    let result = spawnSync(which.sync('psql'), args, {env, stdio: 'inherit'});
+
+    if (result.status !== 0) {
+      let err = new Error('Failed to initialize database');
+      err.failingSpawnReturnValue = result;
+      throw err;
+    }
+  }
+
+  /**
+   * Begin a transaction and return the client which should be used for all
+   * database operations
+   */
+  async beginTx() {
+    let client = await this._getClient();
+    await client.query("BEGIN");
+    return client;
+  }
+
+  /**
+   * Complete a transaction with success
+   */
+  async commitTx(client) {
+    assert(typeof client === 'object');
+    try {
+      await client.query("COMMIT");
+    } finally {
+      client.release();
+    }
+  }
+
+  /**
+   * Complete a transaction with success
+   */
+  async commitTx(client) {
+    assert(typeof client === 'object');
+    try {
+      await client.query("ROLLBACK");
+    } finally {
+      client.release();
+    }
+  }
+
+  /**
+   * Get a client based on one passed in or it that's falsy, create a new one
+   */
+  async _getClient(client) {
+    if (client) {
+      return client;
+    } else {
+      return this._pgpool.connect();
+    }
+  }
+
+  /**
+   * Either insert or update a spot request.  This ensures an atomic insert or
+   * update, but we never learn which one happened.
+   */
+  async upsertSpotRequest({workerType, region, instanceType, id, state, status}, client) {
+    assert(typeof workerType === 'string');
+    assert(typeof region === 'string');
+    assert(typeof instanceType === 'string');
+    assert(typeof id === 'string');
+    assert(typeof state === 'string');
+    assert(typeof status === 'string');
+
+    let text = [
+      "INSERT INTO spotrequests (id, workerType, region, instanceType, state, status)",
+      "VALUES ($1, $2, $3, $4, $5, $6)",
+      "ON CONFLICT (region, id) DO UPDATE SET state = $5, status = $6;"
```

Comment:
so anywhere that we're doing an upsert, is there any value in also putting some guards around updating the state/status only if they are certain conditions to begin with? I haven't read through everything just yet, so didn't know if tehre could be updates done out of order